crates/trie/sparse/src/trie.rs

```diff
@@ -1057,7 +1057,185 @@ impl<P> RevealedSparseTrie<P> {
+                        actual: actual_value.clone(),
+                    });
+                }
+            }
+            Ok(())
+        }
+
+        let mut current = Nibbles::default(); // Start at the root
+
+        // Inclusion proof
+        //
+        // First, do a quick check if the value exists in our values map.
+        // We assume that if there exists a leaf node, then its value will
+        // be in the `values` map.
+        if let Some(actual_value) = self.values.get(path) {
+            // We found the leaf, check if the value matches (if expected value was provided)
+            check_value_match(actual_value, expected_value, path)?;
+            return Ok(LeafLookup::Found);
+        }
+
+        // If the value does not exist in the `values` map, then this means that the leaf either:
+        // - Does not exist in the trie
+        // - Is missing from the witness
+        // We traverse the trie to find the location where this leaf would have been, showing
+        // that it is not in the trie. Or we find a blinded node, showing that the witness is
+        // not complete.
+        while current.len() < path.len() {
+            match self.nodes.get(&current) {
+                Some(SparseNode::Empty) | None => {
+                    // None implies no node is at the current path (even in the full trie)
+                    // Empty node means there is a node at this path and it is "Empty"
+                    return Ok(LeafLookup::NonExistent { diverged_at: current });
+                }
+                Some(&SparseNode::Hash(hash)) => {
+                    // We hit a blinded node - cannot determine if leaf exists
+                    return Err(LeafLookupError::BlindedNode { path: current.clone(), hash });
+                }
+                Some(SparseNode::Leaf { key, .. }) => {
+                    // We found a leaf node before reaching our target depth
+                    let mut full_path = current.clone();
+                    full_path.extend_from_slice_unchecked(key);
+
+                    if &full_path == path {
+                        // This should have been handled by our initial values map check
+                        if let Some(value) = self.values.get(path) {
+                            check_value_match(value, expected_value, path)?;
+                            return Ok(LeafLookup::Found);
+                        }
+                    }
+
+                    // The leaf node's path doesn't match our target path,
+                    // providing an exclusion proof
+                    return Ok(LeafLookup::NonExistent { diverged_at: current });
+                }
+                Some(SparseNode::Extension { key, .. }) => {
+                    // Check if the path matches the extension key prefix
+                    let mut extended_path = current.clone();
+                    extended_path.extend_from_slice_unchecked(key);
```

Comment:
I don't know how performance critical this code is, but cloning it every time only to be able to return the old path in `LeafLookup::NonExistent` feels a bit wasteful, maybe instead we can do `current.extend_from_slice_unchecked(key)` and slice it back in `return Ok(LeafLookup::NonExistent { ... }`?